enhancements/node-artifacts/certificates-and-pull-secrets.md

```diff
@@ -0,0 +1,199 @@
+from these locations OpenShift fails as neither `openshift-api` nor `builder`
+use the credentials provided during the installation. When users configure
+additional certificates for their own registry these are also copied into node's
+filesystem.
+
+### Goals
+
+- Allow users to import image streams from images hosted in any registry provided
+  during or after cluster installation, including `registry.redhat.io`.
+- Allow users to run `builds` based on images hosted in any registry provided
+  during or after cluster installation without providing any extra credentials.
+
+### Non-Goals
+
+- Change the way users manage cluster certificates/pull secrets.
+
+## Proposal
+
+### User Stories
+
+As an OpenShift user
+I want to be able to import image streams from any image registry provided
+during or after the cluster installation
+So that I can use them without manually creating any extra credentials
+
+As an OpenShift user
+I want to be able to use, during builds, images from any registry provided
+during or after the cluster installation
+So that I can use these images as base for my own images
+
+### Implementation Details/Notes/Constraints
+
+These are the node's filesystem paths we are taking into account:
+
+- `/etc/pki/ca-trust/extracted/pem` contains node's certificates.
+- `/var/lib/kubelet/config.json` contains node's pull secrets.
+- `/etc/docker/certs.d` contains node's docker certificates.
+
+#### Image Stream Import
+
+- Mount these three locations as `hostPath` during `openshift-apiserver` deploy
+  under `/node/etc/pki/ca-trust/extracted/pem` for certificates,
+  `/node/var/lib/kubelet/config.json` for pull secrets and 
+  `/node/etc/docker/certs.d` for the docker certificates. Something similar to:
+
+```
+volumeMounts:
+- mountPath: /node/etc/docker/certs.d
+  name: node-docker-certificates
+- mountPath: /node/etc/pki/ca-trust/extracted/pem
+  name: node-certificates
+- mountPath: /node/var/lib/kubelet/config.json
+  name: node-pullsecrets
+volumes:
+- hostPath:
+    path: /etc/docker/certs.d
+    type: Directory
+  name: node-docker-certificates
+- hostPath:
+    path: /var/lib/kubelet/config.json
+    type: File
+  name: node-pullsecrets
+- hostPath:
+    path: /etc/pki/ca-trust/extracted/pem
+    type: Directory
+```
+
+- Copy node's certificates to `anchor` destination inside `openshift-apiserver`
+  pod before running the binary(`/etc/pki/ca-trust/source/anchors/`).
+- Run `update-ca-trust` and just then start the binary.
+- When user requires to import an image stream, `openshift-apiserver` would
+  merge all pull secrets found in `/node/var/lib/kubelet/config.json` with
+  other pull secrets that may exist on the namespace.
+- Hook a `watcher` to observe certificates changes on node's mounted directories,
+  running `update-ca-trust` when something changes.
+
+#### Builds
+
+- `controller-manager` would mount these three locations inside `build` pod 
+  as `hostPath`(similar to what is done for Image Stream Import).
```

Comment:
Are we worried about this content missing on worker nodes? Do we need to mount these as optional? IIRC we have less control over what exists on a worker vs. the master nodes (where the control plane lives).

Reply:
you're mounting a hostpath.  if the hostpath doesn't exist, i believe you'll just get empty content mounted into the pod.  I don't think there's any concept of mounting it as optional.

Reply:
There is no way of configuring `hostPath` as optional. If the file/directory does not exist on the node the pods won't come up:
```
MountVolume.SetUp failed for volume "invalid-path": hostPath type check failed: /invalid is not a directory
```
For sake of completeness: we may [ask](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L526-L558) to _create it if it does not exist_ but imo this does not apply to our problem as, among other reasons, it may bring headaches regarding permissions later down the line.

One option would be to mount a broader scope inside the pods, for instance the `/var` directory.

Reply:
i think we should just determine if it's reasonable to assume the path always exists.  I suspect it is.

Reply:
I would assume that any node which has a kubelet and is running cri-o would have the requisite directories. @umohnani8 and @mrunalp may be in a better position to confirm.

Reply:
This is already sorted out. Nodes running kubelet always have this requisite.